### PR TITLE
refactor http generator output to http_output connector

### PR DESCRIPTION
**Pull request by ekneg54**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -32,6 +33,12 @@
 * handle missing fields in processors via `_handle_missing_fields` from the field_manager
 * add `LogprepMPQueueListener` to outsource logging to a separate process
 * add a single `Queuehandler` to root logger to ensure all logs were handled by `LogprepMPQueueListener`
+* refactor `http_generator` to use a logprep http output connector
+
+### Bugfix
+
+* rename `--thread_count` parameter to `--thread-count` in http generator
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply CHANGELOG.md:40]: sure -> changed

File: logprep/connector/http/output.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        return self._config.password
+
+    @cached_property
+    def _headers(self):
+        return {"Content-Type": "application/x-ndjson; charset=utf-8"}
+
+    @property
+    def statistics(self) -> str:
+        """Return the statistics of this connector as a formatted string."""
+        stats: dict = {}
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply logprep/connector/http/output.py:95]: it is not a good idea, to name variables by its type. only very old C guyes do it ;-)

I return the output of json.dumps directly. So this is fixed.

File: logprep/connector/http/output.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+            default="",
+            converter=lambda x: "" if x is None else x,
+        )
+        """User that is used for the basic auth http request"""
+        password: str = field(
+            validator=validators.instance_of(str),
+            default="",
+            converter=lambda x: "" if x is None else x,
+        )
+        """Password that is used for the basic auth http request"""
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply logprep/connector/http/output.py:74]: yes sure... this is another PR :D

File: logprep/connector/http/output.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+                )
+                stats[key] = int(sample.value)
+        stats = json.dumps(stats, sort_keys=True, indent=4, separators=(",", ": "))
+        return stats
+
+    def store_custom(self, document: dict | tuple | list, target: str) -> None:
+        """
+        Send a batch of events to an endpoint and return the received status code times the number
+        of events.
+        """
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply logprep/connector/http/output.py:117]: fixed

File: logprep/connector/http/output.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    def store(self, document: tuple[str, dict] | dict) -> None:
+        if isinstance(document, tuple):
+            target, document = document
+            target = f"{self._config.target_url}{target}"
+        else:
+            target = self._config.target_url
+        self.store_custom(document, target)
+
+    def store_failed(self, error_message, document_received, document_processed) -> None:
+        self.metrics.number_of_failed_events += 1
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply logprep/connector/http/output.py:129]: done

File: logprep/connector/http/output.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+    def _send_post_request(self, event_target: str, documents: dict | tuple | list) -> dict:
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply logprep/connector/http/output.py:131]: deleted the method signature and use store_custom

File: logprep/generator/http/controller.py [hunk 7]
```diff
@@ -3,74 +3,71 @@
...
-    def _generate_with_main_process(self, output, statistics):
-        for batch in self.input.load():
-            stats = output.send(batch)
-            self._update_statistics(statistics, stats)
-
-    def _update_statistics(self, statistics, new_statistics):
-        statistics.update(new_statistics)
-        if self.use_reporter:
-            self.reporter.update(new_statistics)
+            results = executor.map(self.output.store, self.input.load())
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply logprep/generator/http/controller.py:73]: I think this version is more readable and you have the option to mock the real load generation

File: logprep/generator/http/controller.py [hunk 8]
```diff
@@ -3,74 +3,71 @@
...
     def __init__(self, **kwargs):
         self.config = kwargs
+        self.loghandler = None
+        self._setup_logging()
         self.thread_count: int = kwargs.get("thread_count")
         self.use_reporter: bool = kwargs.get("report")
-        self.log: Logger = logging.getLogger("Generator")
         self.input: Input = Input(self.config)
-        if self.use_reporter:
-            self.reporter = Reporter(args=kwargs)
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply logprep/generator/http/controller.py:31]: deleted the tests and module

File: tests/unit/connector/test_http_output.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+        self.object.metrics.number_of_failed_events = 0
+        self.object.store_custom(123, input_data)
+        assert self.object.metrics.number_of_failed_events == 1, testcase
+
+    @responses.activate
+    def test_store_counts_http_status_codes(self):
+        responses.add(responses.POST, f"{TARGET_URL}/", status=200)
+        responses.add(responses.POST, f"{TARGET_URL}/notfound", status=404)
+        responses.add(responses.POST, f"{TARGET_URL}/transport-error", status=429)
+        responses.add(responses.POST, f"{TARGET_URL}/sucess", status=200)
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply tests/unit/connector/test_http_output.py:119]: fixed

File: logprep/connector/http/output.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+        """Send a post request with given data to the specified endpoint"""
+        document_count = 0
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply logprep/connector/http/output.py]: removed

File: logprep/connector/http/output.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+.. security-best-practice::
+   :title: Http Output Connector - Usage
+
+   This Connector is currently only used in the log generator and does not have a stable interface.
+   Do not use this in production.
+
+.. security-best-practice::
+   :title: Http Output Connector - SSL
+
+   TThis connector does not verify the SSL Context, which could lead to exposing sensitive data.
```
Comment (dtrai2) [inline logprep/connector/http/output.py:37]:
```suggestion
   This connector does not verify the SSL Context, which could lead to exposing sensitive data.
```

**Review by dtrai2 — APPROVED** [review]

**Pull request merged**